Latest 5.12.0 RNs (GW & Dashboard).mdx (#1466)
TykTechnologies/tyk-docs: developer-support/release-notes/gateway.mdx

---

## 5.12 Release Notes 

### 5.12.0 Release Notes 

> #### Release xx February 2026

Reviewer: ```suggestion
#### Release 06 March 2026
```

#### Release Highlights

In this release,
For a comprehensive list of changes, please refer to the detailed [changelog](#Changelog-v5.12.0) below.

#### Breaking Changes

A change has been made to improve security of the legacy *Auth Token + Dynamic mTLS* method for securing access to APIs deployed on Tyk. This removes the option to authenticate using only the auth token and enforces the mTLS handshake.

Previously API clients could authenticate without presenting the client certificate or holding the client's private key.

> For any user relying on that behavior, we have added a new Gateway configuration option: `allow_unsafe_dynamic_mtls_token`. Unless deliberately configured in the config file or environment, this is set to `false` to ensure that Tyk is secure by default.

Reviewer: ```suggestion
A change has been made to improve the security of the legacy **Auth Token + Dynamic mTLS** method for securing access to APIs deployed on Tyk. This removes the option to authenticate using only the auth token and enforces the mTLS handshake.

Previously, API clients could authenticate without presenting the client certificate or holding the client's private key.

For any user relying on that behavior, we have added a new Gateway configuration option: `allow_unsafe_dynamic_mtls_token`. Unless deliberately configured in the config file or environment, this is set to `false` to ensure that Tyk is secure by default.
```

| [Redis](https://redis.io/download/)    | 6.2.x, 7.x, 7.4.x      | 6.2.x, 7.x, 7.4.x      | | 
| [Valkey](https://valkey.io/download/)  | 7.2.x, 8.0.x, 8.1.x    | 7.2.x, 8.0.x, 8.1.x    | | 
| [OpenAPI Specification](https://spec.openapis.org/oas/v3.0.3)| v3.0.x  | v3.0.x | Supported by [Tyk OAS](/api-management/gateway-config-tyk-oas) |

Given the potential time difference between your upgrade and the release of this version, we recommend users verify the ongoing support of third-party dependencies they install, as their status may have changed since the release.

#### Deprecations

There are no deprecations in this patch release.

#### Upgrade instructions
<a id="upgrade-5.12.0"></a>

If you are upgrading to 5.12.0, please follow the detailed [upgrade instructions](#upgrading-tyk).

#### Downloads

- [Docker image to pull](https://hub.docker.com/r/tykio/tyk-gateway/tags?page=&page_size=&ordering=&name=v5.12.0)
  - ```bash
    docker pull tykio/tyk-gateway:v5.12.0
    ``` 
- Helm charts
  - [tyk-charts v5.1.0](/developer-support/release-notes/helm-chart#5-1-0-release-notes)

> 
> Please note that the Tyk Helm Charts are configured to install the LTS version of Tyk Gateway. You will need to modify them to install v5.12.0.
> 

Reviewer: ```suggestion
<Note>
Tyk Helm Charts are configured to install the LTS version of Tyk Gateway. You will need to modify them to install v5.12.0.
</Note>
```

---

## 5.12 Release Notes 

### 5.12.0 Release Notes 

#### Release xx February 2026

#### Release Highlights

> In this release,

Reviewer: Pending

---

## 5.12 Release Notes 

### 5.12.0 Release Notes 

#### Release xx February 2026

#### Release Highlights

In this release,
For a comprehensive list of changes, please refer to the detailed [changelog](#Changelog-v5.12.0) below.

#### Breaking Changes

> There are no breaking changes in this release.

Reviewer 2: ```suggestion
A change has been made to improve security of the legacy *Auth Token + Dynamic mTLS* method for securing access to APIs deployed on Tyk. This removes the option to authenticate using only the auth token and enforces the mTLS handshake.

Previously API clients could authenticate without presenting the client certificate or holding the client's private key.

For any user relying on that behavior, we have added a new Gateway configuration option: `allow_unsafe_dynamic_mtls_token`. Unless deliberately configured in the config file or environment, this is set to `false` to ensure that Tyk is secure by default.
```

</Accordion>

<Accordion title='Comprehensive Certificate Expiry Monitoring for All Gateway Certificate Types'>
Extended certificate expiry monitoring to cover all Gateway certificate types, significantly improving operational visibility and preventing unexpected service disruptions. The Gateway now monitors upstream mTLS certificates used for backend connections and provides startup logging for server and CA certificates. Certificate expiry events now include a `cert_role` field to distinguish between "client" and "upstream" certificates, enabling administrators to route alerts appropriately and understand which part of the system is affected. Upstream certificate monitoring operates at the API level with periodic checking and fires `CertificateExpiringSoon` and `CertificateExpired` events with proper `api_id` attribution when certificates approach or exceed their expiry dates. Server and CA certificates are logged during Gateway startup with expiry information, providing visibility into file-based server certificates, Certificate Store server certificates, and CA certificates used for client verification.
</Accordion>

> <Accordion title='Custom Trace Header Configuration for Legacy Correlation ID Systems'>

Reviewer 3: ```suggestion
<Accordion title='Custom Trace Header Configuration for Custom Correlation ID Systems'>
```

- **Custom-only** (read and write custom headers exclusively)
- **Hybrid** (read custom headers, write standard traceparent)
- **Composite** (read custom headers, write both custom and standard formats)

This flexibility allows gradual migration from legacy correlation systems to standard OpenTelemetry tracing while maintaining trace continuity across the entire request flow. The implementation includes automatic fallback to standard propagators when custom headers are missing or invalid, ensuring robust trace handling in mixed environments. Custom trace headers are validated to conform to OpenTelemetry trace ID standards (32 hex characters) and integrate seamlessly with existing OpenTelemetry exporters and observability platforms, providing organizations with a smooth transition path to modern distributed tracing.
</Accordion>

> <Accordion title='Enhanced 5XX Error Observability in Access Logs'>

Reviewer 3: Let's combine 4xx and 5xx as one entry.

```suggestion
<Accordion title='Enhanced Error Observability in Access Logs'>
```

> Added structured error classification to access logs for 5XX upstream errors to significantly improve troubleshooting capabilities and reduce time-to-resolution for production incidents. Access logs now include detailed error context through new fields:

Reviewer 3: ```suggestion
Added structured error context to access logs for 4XX or 5XX gateway and upstream errors to significantly improve troubleshooting capabilities and reduce time-to-resolution for production incidents. Access logs now include detailed error context through new fields:
```

- `response_flag` (standardized error codes like TLE for TLS certificate expired, UCF for connection refused) 
- `response_code_details` (human-readable error descriptions)
- `error_source` (originating component)

> - `error_target` (upstream address). 

Reviewer 3: ```suggestion
- `error_target` (upstream address). 
- `upstream_status` (HTTP status from upstream).
- `tls_cert_expiry` (TLS certificate expiration date, only present on TLS certificate errors).
- `tls_cert_subject` (TLS certificate subject, only present on TLS certificate errors).
- `circuit_breaker_state` (Circuit breaker state, only present on circuit breaker errors).
```

> The enhancement covers comprehensive error scenarios, including TLS certificate issues (expired, invalid, hostname mismatch), connection problems (refused, timeout, reset), infrastructure failures (DNS resolution, circuit breaker open, no healthy upstreams), and client disconnections. Additional context fields provide specific details such as TLS certificate expiry dates and subjects for certificate-related errors, and circuit breaker states for service protection scenarios. This eliminates the need to cross-reference multiple log sources when investigating 5XX errors, enabling operators to quickly identify whether failures stem from expired certificates, network connectivity issues, or backend service problems directly from access logs.

Reviewer 3: ```suggestion
The `response_flag` helps you identify the root cause of error, including TLS certificate issues (expired, invalid, hostname mismatch), connection problems (refused, timeout, reset), infrastructure failures (DNS resolution, circuit breaker open, no healthy upstreams), rate limiting (RLT), quota exhaustion (QEX), missing authentication headers (AMF), etc. The full list of response flag codes are documented in the [Access Logs](https://tyk.io/docs/api-management/logs#access-logs) section in the Docs. Additional context fields provide specific details such as TLS certificate expiry dates and subjects for certificate-related errors, and circuit breaker states for service protection scenarios. This eliminates the need to cross-reference multiple log sources when investigating API errors, enabling operators to quickly identify whether failures stem from expired certificates, network connectivity issues, or backend service problems directly from access logs.
```

> </Accordion>

Reviewer 3: ```suggestion
```

<Accordion title='Gateway Configuration Dump via Management API for Enhanced Troubleshooting'>

> Added comprehensive configuration inspection endpoints to the Gateway Control API to enable troubleshooting and configuration auditing without requiring SSH access. The new `/config` and `/env` endpoints provide programmatic access to the Gateway's actual runtime configuration, including environment variable mappings and resolved values from multiple configuration sources (file, environment variables, defaults). Platform engineers and support teams can now quickly verify configuration settings, debug deployment issues, and audit security configurations through authenticated API calls. The feature includes intelligent, sensitive data protection, automatically redacting passwords, secrets, and connection strings while preserving configuration structure visibility. Single-field queries support targeted troubleshooting (e.g., `/config?field=storage.host`), and environment variable mappings help clarify configuration precedence and naming conventions. The endpoints reflect live configuration state, including hot-reloaded changes, and maintain consistency with MDCB's configuration inspection format. This enhancement significantly reduces support friction by eliminating manual configuration file sharing and provides automated configuration drift detection capabilities for enterprise deployments. The feature is disabled by default and requires explicit enablement via `enable_config_inspection: true` with proper authentication using the Gateway's Control API secret.

Reviewer 3: ```suggestion
Added configuration inspection endpoints to the Gateway Control API to enable troubleshooting and configuration auditing. The new `/config` and `/env` endpoints provide programmatic access to the Gateway's actual runtime configuration, including environment variable mappings and resolved values from multiple configuration sources (file, environment variables, defaults). Platform engineers and support teams can now quickly verify configuration settings, debug deployment issues, and audit security configurations through authenticated API calls. The feature includes sensitive data protection, automatically redacting passwords, secrets, and connection strings while preserving configuration structure visibility. Both full configuration dump and single-field queries are supported, which help you clarify configuration precedence, naming conventions, and targeted troubleshooting (e.g., `/config?field=storage.host`). This enhancement could be used to significantly reduce support friction by eliminating manual configuration file sharing and supports automated configuration drift detection for enterprise deployments. The feature is disabled by default and requires explicit enablement via `enable_config_inspection: true` with proper authentication using the Gateway's Control API secret.
```

</Accordion>

<Accordion title='Enhanced MDCB `/dataplane` Endpoint with Information on Loaded APIs and Policies'>
The MDCB `/dataplanes` endpoint has been enhanced to provide visibility into your Tyk deployment. This now returns a detailed list of the APIs and policies loaded on each connected data plane gateway. You can retrieve the results for a specific Gateway using the `node_id` query parameter. This update provides a clear picture of what is running on each Gateway, simplifying monitoring and troubleshooting of your deployed Data Planes.

> </Accordion>

Reviewer 3: ```suggestion
<Accordion title='Support Enhanced MDCB `/dataplane` Endpoint with Information on Loaded APIs and Policies'>
Gateway is updated to support the enhanced MDCB `/dataplanes` endpoint. The enhanced MDCB endpoint will provide visibility into your Tyk deployment. This now returns a detailed list of the APIs and policies loaded on each connected data plane gateway. You can retrieve the results for a specific Gateway using the `node_id` query parameter. This update provides a clear picture of what is running on each Gateway, simplifying monitoring and troubleshooting of your deployed Data Planes.
</Accordion>
```

- Support for `exclusiveMinimum` and `exclusiveMaximum` 

Please note the following limitations in this initial release:
- Reusable Path Item Objects and the new `mutualTLS` security scheme are not currently supported.
</Accordion>

<Accordion title='Optimized Data Plane Redis storage for sessions and API definitions'>
This release introduces significant storage optimizations for Data Plane Redis caching, reducing memory consumption while maintaining full backward compatibility:

- *Session Object Optimization*: The Gateway now automatically omits empty and zero-value fields when storing API keys and session objects. This results in up to a 20% reduction in memory usage, with minimal keys now consuming only 500-700 bytes.
- *API Definition Compression*: Added optional Zstd compression for API definitions cached in Redis, achieving up to 75% storage reduction. Compression and decompression occur during Gateway reloads, ensuring zero impact on the request hot path.
- *Configuration*: API definition compression is disabled by default. It can be enabled via the new `storage.enable_api_definition_compression` configuration option.

> - *Security Limit*: To mitigate the risk of decompression bombs, the maximum supported uncompressed size for an API definition is limited to 100MB.

Reviewer 2: ```suggestion
- *Security Limit*: To mitigate the risk of decompression bombs, the maximum combined uncompressed size for the API definitions is limited to 100MB.
```

#### Release Highlights

**OpenAPI Specification 3.1 is now supported**

In this release, we are delighted to bring initial support for OAS 3.1 in Tyk OAS APIs covering:

- Import and validation of OpenAPI 3.1 descriptions using Tyk Dashboard to create Tyk OAS APIs
- OAS 3.1 features:
    - Full JSON Schema Support and $schema keyword
    - The single `example` keyword is deprecated in OAS 3.1
    - `type` can be an array
    - `exclusiveMinimum` and `exclusiveMaximum` keywords

We do not yet have support for all new features. For more details, see the [documentation](/api-management/gateway-config-tyk-oas#openapi-specification-3.1)

**Enhanced OpenTelemetry Tracing and Log Correlation**

In this release, we've significantly improved observability by bridging the gap between logs and distributed traces. When OpenTelemetry is enabled, Tyk Gateway now automatically injects W3C trace IDs into access logs, and both trace and span IDs into application logs. 

This allows your DevOps and SRE teams to seamlessly correlate Gateway operational events with distributed traces across platforms such as Grafana Tempo, Jaeger, and OpenSearch, providing full visibility into the request journey.

> Additionally, we've introduced flexible support for custom trace headers. If your organization uses legacy correlation ID systems (like `X-Correlation-ID`), the Gateway can now recognize these as trace context sources. With multiple propagation modes, you can gradually migrate to standard OpenTelemetry tracing without modifying existing downstream systems.

Reviewer 3: ```suggestion
Additionally, we've introduced flexible support for custom trace headers. If your organization uses custom correlation ID systems (like `X-Correlation-ID`), the Gateway can now recognize these as trace context sources. With multiple propagation modes, you can gradually migrate to standard OpenTelemetry tracing without modifying existing downstream systems.
```
